pkg/actors/internal/placement.go

```diff
@@ -0,0 +1,374 @@
+			}
+			if p.shutdown {
+				break
+			}
+
+			// appHealthFn is the health status of actor service application. This allows placement to update
+			// memberlist and hashing table quickly.
+			if !p.appHealthFn() {
+				// app is unresponsive, close the stream and disconnect from the placement service.
+				// Then Placement will remove this host from the member list.
+				log.Debug("disconnecting from placement service by the unhealthy app.")
+
+				p.closeStream()
+				continue
+			}
+
+			host := v1pb.Host{
+				Name:     p.runtimeHostName,
+				Entities: p.actorTypes,
+				Id:       p.appID,
+				Load:     1, // Not used yet
+				// Port is redundant because Name should include port number
+			}
+
+			err := p.clientStream.Send(&host)
+			if err != nil {
+				diag.DefaultMonitoring.ActorStatusReportFailed("send", "status")
+				log.Debugf("failed to report status to placement service : %v", err)
+			}
+
+			time.Sleep(statusReportHeartbeatInterval)
+		}
+	}()
+}
+
+// Stop shutdowns server stream gracefully.
+func (p *ActorPlacement) Stop() {
+	p.shutdown = true
+	close(p.streamConnectedCh)
+	p.closeStream()
+	p.shutdownConnLoop.Wait()
+}
+
+func (p *ActorPlacement) closeStream() {
+	if p.clientStream != nil {
+		p.clientStream.CloseSend()
+	}
+
+	if p.clientConn != nil {
+		p.clientConn.Close()
+	}
+}
+
+func (p *ActorPlacement) establishStreamConn() (v1pb.Placement_ReportDaprStatusClient, *grpc.ClientConn) {
+	for !p.shutdown {
+		serverAddr := p.serverAddr[p.serverIndex]
+
+		// Stop reconnecting to placement until app is healthy.
+		if !p.appHealthFn() {
+			time.Sleep(placementReconnectInterval)
+			continue
+		}
+
+		log.Infof("try to connect placement service: %s", serverAddr)
+
+		opts, err := dapr_credentials.GetClientOptions(p.clientCert, security.TLSServerName)
+		if err != nil {
+			log.Errorf("failed to establish TLS credentials for actor placement service: %s", err)
+			return nil, nil
+		}
+
+		if diag.DefaultGRPCMonitoring.IsEnabled() {
+			opts = append(
+				opts,
+				grpc.WithUnaryInterceptor(diag.DefaultGRPCMonitoring.UnaryClientInterceptor()))
+		}
+
+		conn, err := grpc.Dial(serverAddr, opts...)
+	NEXT_SERVER:
+		if err != nil {
+			log.Debugf("error connecting to placement service: %v", err)
+			conn.Close()
+			p.serverIndex = (p.serverIndex + 1) % len(p.serverAddr)
+			time.Sleep(placementReconnectInterval)
+			continue
+		}
+
+		client := v1pb.NewPlacementClient(conn)
+		stream, err := client.ReportDaprStatus(context.Background())
+		if err != nil {
+			goto NEXT_SERVER
+		}
+
+		log.Infof("established connection to placement service at %s", serverAddr)
+		return stream, conn
+	}
+
+	return nil, nil
+}
+
+func (p *ActorPlacement) onPlacementOrder(in *v1pb.PlacementOrder) {
+	log.Infof("placement order received: %s", in.Operation)
+	diag.DefaultMonitoring.ActorPlacementTableOperationReceived(in.Operation)
+
+	// lock all incoming calls when an updated table arrives
+	p.operationUpdateLock.Lock()
```

Comment:
The previous placement server code can send multiple table updates. so it is possible to get multiple placement table operation. So it locks for entire operation. however, new change will minimize this table update a lot because new placement behavior is sending one time in the given time. but I would not like to change any existing behavior now. 

Regarding the below TODO comment, we need to redesign entire three stage commit algorithm. actually the current algorithm is more like two stage distributed lock rather than three stage commit. anyhow, we will need to revisit this algorithm instead of fixing now. so I would keep it as is.